src/nvim/os/shell.c

```diff
@@ -309,15 +318,140 @@ static void system_data_cb(Stream *stream, RBuffer *buf, size_t count,
   dbuf->len += nread;
 }
 
+/// Tracks output received for the current executing shell command, and displays
+/// a pulsing "..." when output should be skipped. Tracking depends on the
+/// synchronous/blocking nature of ":!".
+//
+/// Purpose:
+///   1. CTRL-C is more responsive. #1234 #5396
+///   2. Improves performance of :! (UI, esp. TUI, is the bottleneck).
+///   3. Avoids OOM during long-running, spammy :!.
+///
+/// Vim does not need this hack because:
+///   1. :! in terminal-Vim runs in cooked mode, so CTRL-C is caught by the
+///      terminal and raises SIGINT out-of-band.
+///   2. :! in terminal-Vim uses a tty (Nvim uses pipes), so commands
+///      (e.g. `git grep`) may page themselves.
+///
+/// @param size Length of data, used with internal state to decide whether
+///             output should be skipped. size=0 resets the internal state and
+///             returns the previous decision.
+///
+/// @returns true if output should be skipped and pulse was displayed.
+///          Returns the previous decision if size=0.
+static bool out_data_decide_throttle(size_t size)
+{
+  static uint64_t   started     = 0;  // Start time of the current throttle.
+  static size_t     received    = 0;  // Bytes observed since last throttle.
+  static size_t     visit       = 0;  // "Pulse" count of the current throttle.
+  static size_t     max_visits  = 0;
+  static char       pulse_msg[] = { ' ', ' ', ' ', '\0' };
+
+  if (!size) {
+    bool previous_decision = (visit > 0);
+    started = received = visit = 0;
+    max_visits = 20;
+    return previous_decision;
+  }
+
+  received += size;
+  if (received < OUT_DATA_THRESHOLD
+      // Display at least the first chunk of output even if it is big.
+      || (!started && received < size + 1000)) {
+    return false;
+  } else if (!visit) {
+    started = os_hrtime();
+  } else if (visit >= max_visits && size < 256 && max_visits < 999) {
+    // Gobble up small chunks even if we maxed out. Avoids the case where the
+    // final displayed chunk is very tiny.
+    max_visits = visit + 1;
+  } else if (visit >= max_visits) {
+    uint64_t since = os_hrtime() - started;
+    if (since < NS_1_SECOND) {
+      // Adjust max_visits based on the current relative performance.
+      // Each "pulse" period should last >=1 second so that it is perceptible.
+      max_visits = (2 * max_visits);
+    } else {
+      received = visit = 0;
+      return false;
+    }
+  }
```

Comment:
Thanks for that test case. I changed it to timeout after 3 seconds instead of max_visits > 999.

To make it perfect, I guess 2 ring buffers would be needed. Hopefully we're close to "good enough". After looking more closely at the `:!` and `system()` implementations I'm convinced that we could change the implementation of `:!` to *always* use a pty/terminal job. So this throttling logic may not be needed after that (except on Windows, unless we get winpty working).